spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,308 @@
+package org.apache.spark.sql.cassandra
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.Logging
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.cassandra.CassandraSQLRow.CassandraSQLRowReader
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{types, DataFrame, Row, SQLContext}
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnector, ColumnDef}
+import com.datastax.spark.connector.rdd.{ReadConf, ValidRDDType}
+import com.datastax.spark.connector.writer.{WriteConf, SqlRowWriter}
+
+import DataTypeConverter._
+
+/**
+ *  Implements [[BaseRelation]] and [[InsertableRelation]].
+ *  It inserts data into Cassandra table. It also provides some
+ *  helper methods and variables for sub class to use.
+ */
+private[cassandra] abstract class BaseRelationImpl(
+    tableIdent: TableIdent,
+    connector: CassandraConnector,
+    readConf: ReadConf,
+    writeConf: WriteConf,
+    userSpecifiedSchema: Option[StructType],
+    override val sqlContext: SQLContext)
+  extends BaseRelation
+  with InsertableRelation
+  with Serializable
+  with Logging {
+
+  protected[this] val tableDef =
+    sqlContext.getCassandraSchema(tableIdent.cluster.getOrElse("default"))
+      .keyspaceByName(tableIdent.keyspace).tableByName(tableIdent.table)
+
+  override def schema: StructType = {
+    userSpecifiedSchema.getOrElse(StructType(tableDef.allColumns.map(toStructField)))
+  }
+
+  protected[this] val baseRdd =
+    sqlContext.sparkContext.cassandraTable[CassandraSQLRow](
+      tableIdent.keyspace,
+      tableIdent.table)(
+        connector,
+        readConf,
+        implicitly[ClassTag[CassandraSQLRow]],
+        CassandraSQLRowReader,
+        implicitly[ValidRDDType[CassandraSQLRow]])
+
+  override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+    if (overwrite) {
+      val clusterName = tableIdent.cluster.getOrElse(sqlContext.getCluster)
+      new CassandraConnector(sqlContext.getCassandraConnConf(clusterName))
+        .withSessionDo {
+        session => session.execute(s"TRUNCATE ${quoted(tableIdent.keyspace)}.${quoted(tableIdent.table)}")
+      }
+    }
+
+    data.rdd.saveToCassandra(
+      tableIdent.keyspace,
+      tableIdent.table,
+      AllColumns,
+      writeConf)(
+        connector,
+        SqlRowWriter.Factory)
+  }
+
+  override def sizeInBytes: Long = {
+    val keyspace = tableIdent.keyspace
+    val table = tableIdent.table
+    val size = sqlContext.conf.getConf(s"spark.cassandra.$keyspace.$table.size.in.bytes", null)
```

Comment:
This should be first read from the Cassandra system table, and only if unavailable, then taken from the settings.